### ISSUE TITLE
ClojureScript ref lookup with random string matches UUID incorrectly

### ISSUE THREAD

**Issue by timothypratley** [tonsky/datascript#401]
Hi,

This is a weird one; ref lookup in ClojureScript using a string on entities that contain UUIDS:

This is a very basic DB where `:baz/id` is a unique identity

```
(def s "#datascript/DB {:schema {:baz/id   {:db/unique :db.unique/identity}}, :datoms [[1 :baz/fun \"hahaha\" 536870913] [1 :baz/id #uuid \"322a67b0-2834-41f7-9016-c79a2a1b0251\" 536870913] [2 :baz/fun \"hohoho\" 536870914] [2 :baz/id #uuid \"1241a6b1-e61b-4d10-85a3-b24a86e448f3\" 536870914] [3 :baz/fun \"hehehee\" 536870915] [3 :baz/id #uuid \"997bbac7-4e4b-4997-9ce0-2e6072d19799\" 536870915]]}")

(def db2 (cljs.reader/read-string s))
```

Expected: The following should fail to find anything because there are 3 `baz` entities, which all have uuid ids

```
(into {} (d/entity db2 [:baz/id "what?"]))`
=> {:baz/fun "hohoho", :baz/id #uuid "1241a6b1-e61b-4d10-85a3-b24a86e448f3"}
```
Actually: it returns a result. doesn't matter what string you provide, any string gets the same result.

* Using a #uuid the ref lookup behaves correctly
* Does not behave like this in Clojure

**Comment by tonsky** [comment]
Interesting! Unfortunately, I can’t reproduce this. Your example gives me `{}`, as expected.

Are you, by any chance, use transit-cljs anywhere in your system? I remember they had a problem with shipping their own UUIDs that we incompatible with CLJS. But that’s all I can think of.

**Comment by timothypratley** [comment]
Hi @tonsky 


I have created a minimal reproduction here: https://github.com/timothypratley/xy (Please see the readme for more info)

1. The problem manifests when building via Figwheel, but NOT when building/testing via Kaocha!!!
2. I can't seem to reproduce the problem using datoms, only when deserializing, which is probably another good hint that `transit-cljs` is at fault somehow.
3. I think a good theory here is that Figwheel causes a reader to be installed as a side-effect. I am not sure how to check for that. Any tips you have would be very helpful.

Sorry that it took me so long to get back to this.


Sincerely,
Timothy

**Comment by timothypratley** [comment]
O.K. something else interesting I discovered:

If in the file that calls `read-string` I require your library: `[datascript.transit :as dt]`, then it works as expected under figwheel. So my theory is that the #uuid reader is under contention. Note that I'm not using transit at all... but I think it is related because of the reader tag handling.

**Comment by timothypratley** [comment]
And https://github.com/lambdaisland/kaocha-cljs/search?q=uuid

Kaocha is extending doing some special stuff here which is why the problem doesn't show up there.

**Comment by timothypratley** [comment]
It seems to me that if DataScript itself followed Kaocha's example here and added:

````
(extend-type UUID
  IEquiv
  (-equiv [this other]
    (cond
      (instance? UUID other)
      (identical? (.-uuid this) (.-uuid other))

      (instance? ty/UUID other)
      (identical? (.-uuid this) (.toString other))

      :else false)))

(extend-protocol IComparable
  UUID
  (-compare [this other]
    (if (or (instance? UUID other)
            (instance? ty/UUID other))
      (compare (.toString this) (.toString other))
      (throw (js/Error. (str "Cannot compare " this " to " other)))))
  ty/UUID
  (-compare [this other]
    (if (or (instance? UUID other)
            (instance? ty/UUID other))
      (compare (.toString this) (.toString other))
      (throw (js/Error. (str "Cannot compare " this " to " other))))))
```

Then it would allow UUIDs to be used as identities.
What do you think?
I'm happy to try that out tomorrow and submit a PR if this seems reasonable.

**Comment by tonsky** [comment]
Ok I think I’ve found the root cause. In CLJS UUID/compare used not to check for second argument type:

https://github.com/clojure/clojurescript/blob/r1.10.773/src/main/cljs/cljs/core.cljs#L11295-L11297

In that particular case, `.-uuid` was `null`, and `good.array.defaultCompare` treats everything as being equal to null:

https://github.com/google/closure-library/blob/1224c4cf9d66fd903f3098a262d5763782e99418/closure/goog/array/array.js#L1344-L1346

```
"abc" < null
// => false
"abc" > null
// => false
```

This behavior was fixed in 1.10.844: CLJS-3130: UUID compares equal to other values

Issue: https://clojure.atlassian.net/browse/CLJS-3130
Commit: https://github.com/clojure/clojurescript/commit/25c9587d6ec37d4aacb291bb2fec8b46d5b80a7d

If you update your CLJS version, your test passes. Hopefully this resolves your issue

**Comment by timothypratley** [comment]
Oh wow, nice detective work :)
Yes, that solves my problem.
Thank you very much!!!
